File: algorithm_library/python/techniques/imos.py
from itertools import accumulate

# 二次元のimos法
class imos:
  def __init__(self, H : int, W : int):
    self.grid = [[0]*W for _ in range(H)]
    
  def add(self, i0: int, j0: int, i1: int, j1: int, increment: int) -> None:
    """
    左上を(i0,j0), 右下を(i1,j1)とする長方形領域に increment 加算する
    
    grid[i0][j0], grid[i0][j0+1], ..., grid[i0][j1]
    grid[i0+1][j0], grid[i0+1][j0+1], ..., grid[i0+1][j1]
    ...
    grid[i1][j0], grid[i1][j0+1], ..., grid[i1][j1]
    """
    assert 0<=i0<=i1<len(self.grid) and 0<=j0<=j1<len(self.grid[0])
    
    self.grid[i0][j0] += increment
    if i1+1<len(self.grid): self.grid[i1+1][j0] -= increment
    if j1+1<len(self.grid[0]): self.grid[i0][j1+1] -= increment
    if i1+1<len(self.grid) and j1+1<len(self.grid[0]):
      self.grid[i1+1][j1+1] += increment
      
  # 二次元累積和を取る
  def cross_accumulate(self):
    self.flag = True
    for i in range(len(self.grid)):
      self.grid[i] = list(accumulate(self.grid[i]))
    for j in range(len(self.grid[0])):
      for i in range(1,len(self.grid)):
        self.grid[i][j] += self.grid[i-1][j]
    return self.grid

File: algorithm_library/cpp-from-py/techniques/imos.cpp
#include <bits/stdc++.h>
using namespace std;

struct imos {
    vector<vector<long long>> grid;

    imos(int H, int W) : grid(H, vector<long long>(W, 0)) {}

    void add(int i0, int j0, int i1, int j1, long long increment) {
        assert(0 <= i0 && i0 <= i1 && i1 < (int)grid.size());
        assert(0 <= j0 && j0 <= j1 && j1 < (int)grid[0].size());
        grid[i0][j0] += increment;
        if (i1 + 1 < (int)grid.size()) grid[i1 + 1][j0] -= increment;
        if (j1 + 1 < (int)grid[0].size()) grid[i0][j1 + 1] -= increment;
        if (i1 + 1 < (int)grid.size() && j1 + 1 < (int)grid[0].size()) grid[i1 + 1][j1 + 1] += increment;
    }

    vector<vector<long long>> cross_accumulate() {
        for (auto& row : grid) {
            for (int j = 1; j < (int)row.size(); j++) row[j] += row[j - 1];
        }
        for (int j = 0; j < (int)grid[0].size(); j++) {
            for (int i = 1; i < (int)grid.size(); i++) grid[i][j] += grid[i - 1][j];
        }
        return grid;
    }
};
